karapace/schema_registry_apis.py

```diff
@@ -395,16 +420,39 @@ async def subjects_list(self, content_type):
         subjects_list = [key for key, val in self.ksr.subjects.items() if self.ksr.get_schemas(key)]
         self.r(subjects_list, content_type, status=HTTPStatus.OK)
 
-    async def subject_delete(self, content_type, *, subject):
-        self._subject_get(subject, content_type)
-        version_list = list(self.ksr.get_schemas(subject))
+    async def subject_delete(self, content_type, *, subject, request: HTTPRequest):
```

Comment:
There are lots of copies going and iteration over the entries here.

- `_subject_get` calls `get_schemas` which iterates on all entries to filter the deleted schemas, and returns a new dict
- `_subject_get` makes a copy of the result above (@juha-aiven [suggestion](https://github.com/aiven/karapace/pull/210/files#r627156298) would eliminate this second copy, since it would always be a new dict)
- the check for soft deletes below iterate over all the schemas again
- then `get_schemas` is called a second time, which may make another copy, and then a new list of the keys is copied out
- and then there is another copy in the loop which actually send the hard delete message

It would be nice to improve this.

Reply:
This method needs synchronization with the `subject_post`, if the whole subject is being deleted, a new schema must not be registered.

Reply:
Fixed second reading of `get_schemas` as it was not needed. When working on this I had a bit hard time to understand original code with `_subject_get` and `ksr.get_schemas` both going into ksr's internals.

About locking for me it looks there has not been any locking to synchronize earlier for posting or deleting. Only lock is for id reservation. So if this locking is an issue, should it be addressed separately for good as I fail to see that this PR would introduce any new synchronization issues?

Reply:
> About locking for me it looks there has not been any locking to synchronize earlier for posting or deleting. Only lock is for id reservation. So if this locking is an issue, should it be addressed separately for good as I fail to see that this PR would introduce any new synchronization issues?

This PR is not introducing a new issue, but it is touching the code that needs and lacks synchronization. For me the important thing is to fix the issue, I would prefer to do it now instead of delaying it.